Add image attachment support to federated comments

obenland opened the pull request:
Fixes #1851.

## Proposed changes:
* Refactored image processing methods from Post transformer to Base class for code reuse.
* Added `get_attachment` method to Comment transformer to extract and convert HTML images to ActivityStreams attachments.
* Maintained backwards compatibility with deprecated wrapper methods in Post class.
* Added comprehensive test coverage for comment image attachment functionality.

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:

* Create a test post and navigate to its comments section.
* Add a comment containing HTML image tags: `This is a comment with an image: <img src="[local-image-url]" alt="Test image" />`.
* Verify the comment's ActivityStreams JSON representation includes both content and attachment fields.
* Test with external images to confirm they are ignored for security.
* Run `npm run env-test -- --filter=test_comment` to execute all comment-related tests.
* Verify backwards compatibility by testing any code that uses the deprecated `wp_attachment_to_activity_attachment` method.

### Changelog entry

- [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

- [x] Minor

#### Type

- [x] Added - for new features

#### Message

Add image attachment support to federated comments - HTML images in comment content now include proper ActivityStreams attachment fields.

</details>

obenland committed:
Add image attachment support to federated comments

Refactored image processing methods to Base transformer class for code reuse:
- Moved `get_classic_editor_image_embeds` from Post to Base class
- Added new `transform_attachment` method to Base class
- Moved `get_wordpress_attachment` to Base class
- Maintained backwards compatibility with deprecated wrappers in Post class

Added `get_attachment` method to Comment transformer that:
- Extracts HTML `<img>` tags from comment content
- Converts local WordPress images to ActivityStreams attachments
- Ignores external images for security
- Respects max image attachment limits
- Uses shared transformation logic for consistency

This enables federated comments with embedded images to include proper
ActivityStreams attachment fields in their JSON representation.

Fixes #1851

matticbot committed:
Add changelog

obenland committed:
misc fixes

obenland committed:
Merge branch 'feature/federated-comments-image-attachments' of https://github.com/Automattic/wordpress-activitypub into feature/federated-comments-image-attachments

obenland committed:
Fix deprecation warnings in Post transformer tests

- Updated Post transformer get_attachment method to use Base class transform_attachment directly
- Renamed get_wordpress_attachment to get_attachment_image_src in Base class to avoid method override conflicts
- Updated deprecated wrapper method references to point to new method names
- All tests now pass without unexpected deprecation warnings

obenland committed:
Backslashit

obenland committed:
Fix incorrect array_intersect_key usage and refactor duplicate removal

- Replaced incorrect array_intersect_key usage with proper array_filter logic
- Added filter_unique_attachments method to Base transformer class
- Updated Post and Comment transformers to use shared filtering method
- Eliminates duplicate code and addresses Copilot feedback

The previous code incorrectly used array_intersect_key with unique IDs as values
instead of keys. The new approach properly filters duplicates while maintaining
array structure and order.

obenland committed:
Use get_attachment_image_src for image filter

Replaces get_wordpress_attachment with get_attachment_image_src in the 'activitypub_get_image' filter to ensure correct image source retrieval.

obenland committed:
Fix phpcs

Simplified conditional checks for video dimensions and improved the filter_unique_attachments method for clarity and correctness in class-base.php. Updated comments for better readability and consistency in class-comment.php and related tests.

obenland committed:
Cleanup

pfefferle commented on includes/transformer/class-base.php:
```diff
@@ -381,4 +382,228 @@ protected function get_mentions() {
+	 *
+	 * @return array The updated media array with found images.
+	 */
+	protected function get_classic_editor_image_embeds( $media, $max_images, $content ) {
```
shouldn't we use a more generic name then? something like `extract_images_from_html`?

obenland replied:
Fixed in [d0fbd42](https://github.com/Automattic/wordpress-activitypub/pull/1996/commits/d0fbd42c4b4eb2de82241028582564ab33858a06).

obenland committed:
Merge branch 'trunk' into feature/federated-comments-image-attachments

obenland committed:
Rename get_classic_editor_image_embeds to parse_html_images

Updated method name to better reflect its generic usage for parsing HTML
content for image tags. The new name is more concise and suitable for
its extended use across both Post and Comment transformers.

Addresses feedback on PR #1996 requesting a more generic method name.

obenland committed:
Merge branch 'trunk' into feature/federated-comments-image-attachments

pfefferle reviewed (approved):
nice one! ♥️
